nni/mutable/_notimplemented.py

```diff
@@ -0,0 +1,70 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+# pylint: skip-file
+# type: ignore
+
+"""Several planned features that are under discussion / not implemented yet."""
+
+from __future__ import annotations
+
+from .annotation import MutableAnnotation
+from .mutable import LabeledMutable, MutableSymbol, Discrete, Continuous
+
+
+class Permutation(MutableSymbol):
+    """Get a permutation of several values.
+    Not implemented. Kept as a placeholder.
+    """
+
+
+class RandomInteger(Discrete[int]):
```

Comment:
I want to change `randint` to default to ordered integers.
The old `randint` is nearly meaningless, as different tuners treat it differently.
`quniform` is not a perfect substitute and sometimes causes problems. For example `list_[quniform(0, 10, 1)]` does not work.
If someone wants unordered integers, they can simply use `choice(range(a, b))`.

Reply:
I'm actually fine with both: either making randint a special case of Discrete, or making it an independent MutableSymbol.

The problem here is, however, what if the algorithm doesn't want to handle randint? In the design of `simplify`, there is a `is_leaf` callable, that supports to expand some mutables where you don't want to handle. In this case, you can return false when checking `is_leaf` for randint. And then what?

Option 1: it becomes a discrete.
Option 2: it becomes a quniform.
Option 3: it becomes a `NonNegativeRandomInteger` (see below). This is actually kept for those who doesn't even want to handle Discrete (i.e., don't care about the values, only indices matter.

If you think about this a little bit further, you will notice that `Continuous` are not that easy to handle. That's where `UnitUniform` comes from (and that's option 4).

My point here is that, the decision here is actually about the simplify-"fallback" of mutables. If the algorithm can handle this kind of mutable, it should never be a problem. If it can't, we need to have a backup plan to make their lives easier. And, everything will eventually fallback to `UnitUniform`.

Finally, let's vote for `RandomInteger -> Discrete -> NonNegativeRandomInteger -> UnitUniform`, or `RandomInteger -> NonNegativeRandomInteger -> UnitUniform` or `RandomInteger -> Continous -> UnitUniform`, or `RandomInteger -> UnitUniform`.

Reply:
I have found my original plan, it was `randint(low, high, categorical=False)`.

`UnitUniform` is bad in HPO because it does not support update search space. I know I used it in `ParameterSpec`. That's a mistake.

So,
  - RandomInteger (ordered) -> Continuous
  - RandomInteger (unordered) -> Discrete -> Continuous (one hot)

Reply:
Firstly I don't like utility functions here because it's notoriously hard to extend. When `simplify()` is a member function, all mutables can have the same interface (including the NAS-specific modules and model space). If `simplify()` is a util function, how should I do that?

Secondly Continuous is definitely not the final fallback layer because of its complexity right now. The algorithm needs to check all mu, sigma, low, high, quant, and optionally customized distribution to make sure that they have handled all the cases.
I understand algorithms might have preferences over what the fallback of Continuous should be (uniform or normal), but when `is_leaf` returns false for Continuous, the algorithm have already made the claim that Continuous is not supported. Continuous simplifies itself into uniform for the convenience of **most algorithms**. If you want to handle normal, you should've done it on the Continuous layer and never simplified it.
The position of `simpfily()` is that it makes writing algorithms easier. You can always choose NOT using it.

Adding `categorical=False` to randint might be a good idea, because it belongs to the semantics of mutable itself.